python/ray/serve/autoscaling_metrics.py

```diff
@@ -0,0 +1,98 @@
+import threading
+import bisect
+from collections import defaultdict
+import time
+from typing import Callable, DefaultDict, Dict, List, Optional
+from dataclasses import dataclass, field
+
+
+def start_metrics_pusher(interval_s: float,
+                         collection_callback: Callable[[], Dict[str, float]],
+                         controller_handle):
+    """Start a background thread to push metrics to controller.
+
+    Args:
+        interval_s(float): the push interval.
+        collection_callback: a callable that returns the metric data points to
+          be sent over the the controller.
+        controller_handle: actor handle to Serve controller.
+    """
+
+    def send_once():
+        data = collection_callback()
+        # TODO(simon): maybe wait for ack or handle controller failure?
+        controller_handle.record_autoscaling_metrics.remote(
+            data=data, send_timestamp=time.time())
+
+    def send_forever():
+        while True:
+            start = time.time()
+            send_once()
+            duration_s = time.time() - start
+            remaining_time = interval_s - duration_s
+            if remaining_time > 0:
+                time.sleep(remaining_time)
+
+    timer = threading.Thread(target=send_forever)
+    timer.setDaemon(True)
+    timer.start()
+
+
+@dataclass(order=True)
+class TimeStampedValue:
+    timestamp: float
+    value: float = field(compare=False)
+
+
+class InMemoryMetricsStore:
+    """A very simple, in memory time series database"""
+
+    def __init__(self):
+        self.data: DefaultDict[str, List[TimeStampedValue]] = defaultdict(list)
```

Comment:
do we have any flush / TTL mechanism here to prevent accumulating python objects over time ? We might also save quite some space by using lightweight data objects with _slots_ that stores value in fixed size list.